### PR TITLE
Add ability to load externally generated satellite data files

### PR DESCRIPTION
**Pull request by simonrp84**
In preparation for MTG/FCI, this PR adds the ability to load satellite data (channels + angles + geolocation + ancillary attributes) from a single netCDF4 file.

In the driver file, instead of specifying a sensor name on the first line, specify `PYTHON`.
I have only tested this with SEVIRI and FCI, but it should work more widely. Unsure if dual view will work, however.

Python scripts to generate these netCDF4 input files is available in a separate repo. I'll link to it once I have finalised and tested it.

### CODE REVIEW

File: pre_processing/setup.F90 [hunk 1]
```diff
@@ -1087,6 +1087,142 @@ subroutine setup_modis(args, channel_ids_user, channel_info, verbose)
...
+   end if
+
+   ! Read platform and sensor names.
+   call ncdf_get_single_attribute_str(fid, 'platform', args%platform)
+   call ncdf_get_single_attribute_str(fid, 'sensor', args%sensor)
+
+   if (verbose) write(*,*) "Satellite is: ", args%platform
+
+   ! Get channel attributes
+   call ncdf_get_arr_attribute_real(fid, 'all_channel_wl_abs', all_channel_wl_abs)
```
Comment (adamcpovey) [inline pre_processing/setup.F90:1177]: This isn't a now problem, but ORAC needs a better way of attaching metadata to channels. For this reader, these are a lot of arrays to write which the users needs to understand the arcane meanings of. For the existing readers, there is a lot of hard-coded data.

File: pre_processing/setup.F90 [hunk 2]
```diff
@@ -2162,6 +2301,8 @@ subroutine setup_imager(args, opts, source_attributes, channel_info, verbose)
 
    end select
 
+   !stop
```
Comment (adamcpovey) [inline pre_processing/setup.F90:2304]: This appears to be debugging code.

File: pre_processing/setup.F90 [hunk 3]
```diff
@@ -2055,7 +2191,7 @@ subroutine setup_imager(args, opts, source_attributes, channel_info, verbose)
    implicit none
 
    type(setup_args_t),        intent(inout) :: args
-   type(preproc_opts_t),      intent(in)    :: opts
+   type(preproc_opts_t),      intent(inout) :: opts
```
Comment (adamcpovey) [inline pre_processing/setup.F90:2194]: I don't think is right. `opts%channel_ids` is an array read from the driver file where the user states what channels they want to preprocess, which can be some subset of those available in the file. You're overriding that functionality for Python reads and Python reads only.

I think you can allow the normal behaviour by using `common_setup` and using the read from the Python file in the same way that the hard-coded arrays are used for the other instruments.

**Review by adamcpovey — CHANGES_REQUESTED** [review]
Seems to be a straightforward change, though I have a lot of minor questions and two big ones. (I will always push for subsetting as if I can't subset a sensor, I basically won't run it. The ability to run one pixel is too useful for debugging.)

File: pre_processing/cloud_typing_pavolonis.F90 [hunk 4]
```diff
@@ -2076,8 +2099,9 @@ function plank_inv(input_platform, T)
         2707.560, 0.999085, 0.58063, 5.123, & ! viirs SoumiNpp/NOAA20
         2673.797, 0.994884, 2.19600, 5.064, & ! slstr Sentinel-3
         2692.112, 0.996369, 2.64763, 5.093, & ! agri fengyun-4
+        2631.579, 0.994040, 1.94224, 4.688, & ! MTG-I1 FCI
         2670.000, 0.998000, 1.75000, 5.000  & ! default
-        /), (/ 4, 25 /))
+        /), (/ 4, 27 /))
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/cloud_typing_pavolonis.F90:2104]: No, there are 27 items in the array. It was the wrong shape before.

File: pre_processing/orac_preproc.F90 [hunk 5]
```diff
@@ -709,8 +713,9 @@ subroutine orac_preproc(mytask, ntasks, lower_bound, upper_bound, driver_path_fi
 
    if (granule%startx.ge.1 .and. granule%endx.ge.1 .and. &
         granule%starty.ge.1 .and. granule%endy.ge.1) then
-      if ( trim(adjustl(granule%sensor)) .eq. 'VIIRSI' .or. &
-           trim(adjustl(granule%sensor)) .eq. 'VIIRSM') then
+      if ( trim(adjustl(granule%sensor_rdr)) .eq. 'VIIRSI' .or. &
+           trim(adjustl(granule%sensor_rdr)) .eq. 'VIIRSM' .or. &
+           trim(adjustl(granule%sensor_rdr)) .eq. 'PYTHON') then
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/orac_preproc.F90:718]: That is something for a separate PR - it's beyond the scope of this one, especially as the MTG-I1 data can be subsetted before it's loaded into ORAC.

File: pre_processing/read_python.F90 [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+   integer :: nc
+ 
+   if (verbose) write(*,*) '<<<<<<<<<<<<<<< read_python()'
+   call ncdf_open(ncid, infile, 'read_agri_data()')
+
+   ! Read the channel data
+   call ncdf_read_array(ncid, "channel_data", imager_measurements%data)
+   where(imager_measurements%data(:, :,:)  .gt. 500) &
+      imager_measurements%data(:, :,:) = sreal_fill_value
+   where(imager_measurements%data(:, :,:)  .lt. 0) &
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/read_python.F90:64]: You mean a channel value of 0? It's a valid value for most instruments and must not be used as a fill value.
That said, it wouldn't surprise me if some sat is using it as a fill value, but badly-made sat data isn't a reason to exclude 0 values.

File: pre_processing/read_python.F90 [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      imager_measurements%data(:, :,:) = sreal_fill_value
+
+   ! Read the geolocation data
+   call ncdf_read_array(ncid, "latitude", imager_geolocation%latitude)
+   where(imager_geolocation%latitude(:,:)  .gt. 100) &
+   imager_geolocation%latitude(:,:) = sreal_fill_value
+   where(imager_geolocation%latitude(:,:)  .lt. -100) &
+   imager_geolocation%latitude(:,:) = sreal_fill_value
+   call ncdf_read_array(ncid, "longitude", imager_geolocation%longitude)
+   where(imager_geolocation%longitude(:,:)  .gt. 200) &
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/read_python.F90:74]: Yes. The standard is -180 -> 180 and satpy follows that standard.

File: pre_processing/read_python.F90 [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+   ! Read the channel data
+   call ncdf_read_array(ncid, "channel_data", imager_measurements%data)
+   where(imager_measurements%data(:, :,:)  .gt. 500) &
+      imager_measurements%data(:, :,:) = sreal_fill_value
+   where(imager_measurements%data(:, :,:)  .lt. 0) &
+      imager_measurements%data(:, :,:) = sreal_fill_value
+
+   ! Read the geolocation data
+   call ncdf_read_array(ncid, "latitude", imager_geolocation%latitude)
+   where(imager_geolocation%latitude(:,:)  .gt. 100) &
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/read_python.F90:69]: Satpy deals with this in the case that it computes the geolocation internally, but some satellites supply the geoloc themselves and in that case this is a useful catch.

File: pre_processing/read_python.F90 [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+   call ncdf_read_array(ncid, "satellite_zenith_angle", imager_angles%satzen)
+   where(imager_angles%satzen(:, :,:)  .lt. -200) &
+      imager_angles%satzen(:, :,:) = sreal_fill_value
+
+   call ncdf_read_array(ncid, "satellite_azimuth_angle", imager_angles%satazi)
+   where(imager_angles%satazi(:, :,:)  .lt. -200) &
+      imager_angles%satazi(:, :,:) = sreal_fill_value
+
+   ! Read per-pixel time information
+   call ncdf_read_array(ncid, "time_data", imager_time%time)
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/read_python.F90:101]: This is a larger problem than just this reader, however, and needs to be dealt with separately.

File: pre_processing/setup.F90 (hunk 1 above)
Comment (simonrp84) [reply pre_processing/setup.F90:1177]: Point is that users don't need to write them.

File: pre_processing/read_python.F90 [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+   call ncdf_read_array(ncid, "channel_data", imager_measurements%data)
+   where(imager_measurements%data(:, :,:)  .gt. 500) &
+      imager_measurements%data(:, :,:) = sreal_fill_value
+   where(imager_measurements%data(:, :,:)  .lt. 0) &
+      imager_measurements%data(:, :,:) = sreal_fill_value
+
+   ! Read the geolocation data
+   call ncdf_read_array(ncid, "latitude", imager_geolocation%latitude)
+   where(imager_geolocation%latitude(:,:)  .gt. 100) &
+   imager_geolocation%latitude(:,:) = sreal_fill_value
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/read_python.F90:70]: Fixed.

File: pre_processing/setup.F90 (hunk 2 above)
Comment (simonrp84) [reply pre_processing/setup.F90]: Fixed.

File: pre_processing/allocate_imager_structures.F90 [hunk 11]
```diff
@@ -61,6 +61,8 @@ subroutine allocate_imager_structures(imager_geolocation,imager_angles, &
    type(imager_cloud_t),        intent(out)   :: imager_cloud
    type(channel_info_t),        intent(in)    :: channel_info
 
+   print*,imager_geolocation%startx,imager_geolocation%endx,1,imager_geolocation%ny,1,channel_info%nchannels_total
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/allocate_imager_structures.F90]: Fixed.

File: pre_processing/set_ecmwf.F90 [hunk 12]
```diff
@@ -378,6 +378,7 @@ subroutine determine_jasmin_filenames_era5(nwp_fnames, cyear, cmonth, cday, chou
     ! Now check if we can find files matching this pattern. If not, we
     ! redefine filenase for the provisional ERA5 data record and try
     ! again
+    print*,trim(adjustl(ml_dir))//'q.nc'
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply pre_processing/set_ecmwf.F90]: Fixed.

File: pre_processing/setup.F90 (hunk 3 above)
Comment (simonrp84) [reply pre_processing/setup.F90:2194]: From memory this was the only way I could think of to get it working. Changes welcome.

**Review by adamcpovey — APPROVED** [review]

**Comment by simonrp84** [comment]
Also, I want to upload the python code associated with making the external files but I don't think that should go in this repo. Rather my preference would be to create a separate repo in `orac-cc` for this. Anyone know how I do that?

**Comment by adamcpovey** [comment]
Don has the password for the orac-cc account. I assume you want a separate repo because you're hoping people will use the code for purposes other than ORAC?

**Comment by simonrp84** [comment]
No, I don't expect anyone else to use it. Just want it in a separate repo as the main ORAC repo is already over-cluttered with different tools and sections.

**Pull request merged**